File: python/tests/euler_unit_test.py
import pytest

from euler import explicit_euler_integrate_one_step, explicit_euler_integrate


def test_calls_function():
    has_been_called = False

    def f(x):
        nonlocal has_been_called
        has_been_called = True
        return 5.0

    explicit_euler_integrate_one_step(f, 0, 0)

    assert has_been_called == True


def const_function(x):
    return 5.0


def test_returns_zero_for_delta_zero():
    assert 1.0 == explicit_euler_integrate_one_step(const_function, 1.0, 0.0)


def test_returns_correct_value_for_delta_one():
    assert 5.0 == explicit_euler_integrate_one_step(const_function, 0.0, 1.0)


def test_returns_correct_value_for_delta_two():
    assert 10.0 == explicit_euler_integrate_one_step(const_function, 0.0, 2.0)


def test_returns_correct_value_for_delta_00001():
    assert 5.0 * 0.00001 == explicit_euler_integrate_one_step(const_function, 0.0, 0.00001)


@pytest.mark.parametrize("delta, expected_size", [(1.5, 1),  # only start point contained
                                                  (1.0, 2),  # start and end point
                                                  (0.5, 3),  # start, end and one midpoint
                                                  (0.1, 11)])  # start, end and 9 points contained
def test_euler_returned_map_size(delta, expected_size):
    values = explicit_euler_integrate(const_function, 0.0, delta, 0.0, 1.0)
    assert len(values) == expected_size


@pytest.mark.parametrize("delta", [1.0, 0.5, 0.1])
def test_euler_returned_map_contains_trange_start(delta):
    values = explicit_euler_integrate(const_function, 0.0, delta, 0.0, 1.0)
    assert 0.0 in values.keys()


@pytest.mark.parametrize("delta", [1.0, 0.5])
def test_euler_returned_map_contains_trange_end(delta):
    values = explicit_euler_integrate(const_function, 0.0, delta, 0.0, 1.0)
    assert 1.0 in values.keys()


@pytest.mark.parametrize("delta", [1.0, 0.5])
def test_euler_returned_map_contains_value_for_one_integration_point(delta):
    values = explicit_euler_integrate(const_function, 0.0, delta, 0.0, 1.0)

    value = values[1.0]
    deviation = value - const_function(1.0)
    assert deviation < 1e-5

File: cpp/tests/euler_unit_test.cpp
#include "euler.hpp"
#include <fstream>
#include <gtest/gtest.h>
#include <iostream>

auto const_func = [](double /*v*/) { return 5.0; };

TEST(OneStepTest, CallsFunction)
{
    bool has_been_called { false };
    explicit_euler_integrate_one_step<double>(
        [&has_been_called](double /*unused*/) {
            has_been_called = true;
            return 0.0;
        },
        0.0, 0.0);
    ASSERT_TRUE(has_been_called);
}

TEST(OneStepTest, ReturnsY0ForDeltaZero)
{
    ASSERT_EQ(explicit_euler_integrate_one_step<double>(const_func, 1.0, 0.0), 1.0);
}

TEST(OneStepTest, ReturnsCorrectValueForDelta1)
{
    ASSERT_EQ(5.0, explicit_euler_integrate_one_step<double>(const_func, 0.0, 1.0));
}

TEST(OneStepTest, ReturnsCorrectValueForDelta2)
{
    ASSERT_EQ(10.0, explicit_euler_integrate_one_step<double>(const_func, 0.0, 2.0));
}
TEST(OneStepTest, ReturnsCorrectValueForDelta00001)
{
    ASSERT_EQ(00001 * 5.0, explicit_euler_integrate_one_step<double>(const_func, 0.0, 00001));
}

class EulerUnitTestReturnedMapSizeTestFixture
    : public ::testing::TestWithParam<std::tuple<double, std::size_t>> {
};

// integrate from 0 to 1 and check that for a given delta the correct amount of points is contained.
TEST_P(EulerUnitTestReturnedMapSizeTestFixture, ReturnedMapContainsCorrectAmountOfEntries)
{
    auto const delta = std::get<0>(GetParam());
    auto const expected_size = std::get<1>(GetParam());
    auto values = explicit_euler_integrate<double>(const_func, 0.0, delta, 0.0, 1.0);
    ASSERT_EQ(values.size(), expected_size);
}

INSTANTIATE_TEST_SUITE_P(EulerUnitTestReturnedMapSizeTest, EulerUnitTestReturnedMapSizeTestFixture,
    ::testing::Values(
        std::make_tuple<double, std::size_t>(1.5, 1), // only the start point contained
        std::make_tuple<double, std::size_t>(1.0, 2), // start point and end point contained
        std::make_tuple<double, std::size_t>(0.5, 3), // start, end and one midpoint contained
        std::make_tuple<double, std::size_t>(0.1, 11))); // start, end and 9 points contained

class EulerUnitTestReturnedMapValuesTestFixture : public ::testing::TestWithParam<double> {
};

TEST_P(EulerUnitTestReturnedMapValuesTestFixture, ReturnedMapContainsTRangeStart)
{
    auto const delta = GetParam();
    auto values = explicit_euler_integrate<double>(const_func, 0.0, delta, 0.0, 1.0);
    ASSERT_DOUBLE_EQ(values.cbegin()->first, 0.0);
}
TEST_P(EulerUnitTestReturnedMapValuesTestFixture, ReturnedMapContainsTRangeEnd)
{
    auto const delta = GetParam();
    auto values = explicit_euler_integrate<double>(const_func, 0.0, delta, 0.0, 1.0);
    ASSERT_DOUBLE_EQ((--values.cend())->first, 1.0);
}

TEST_P(EulerUnitTestReturnedMapValuesTestFixture,
    ReturnedMapContainsCorrectValueForOneIterationPointAndConstantFunction)
{
    auto const delta = GetParam();
    auto values = explicit_euler_integrate<double>(const_func, 0.0, delta, 0.0, 1.0);

    double value = (--values.cend())->second;
    double const deviation = value - const_func(1);
    ASSERT_LT(deviation, 1e-5);
}

INSTANTIATE_TEST_SUITE_P(EulerUnitTestReturnedMapValuesTest,
    EulerUnitTestReturnedMapValuesTestFixture, ::testing::Values(1.0, 0.5, 0.1));
